components/p3a/message_manager.h

```diff
@@ -0,0 +1,136 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_P3A_MESSAGE_MANAGER_H_
+#define BRAVE_COMPONENTS_P3A_MESSAGE_MANAGER_H_
+
+#include <memory>
+#include <string>
+
+#include "base/containers/flat_map.h"
+#include "base/functional/callback.h"
+#include "base/memory/raw_ptr.h"
+#include "base/memory/ref_counted.h"
+#include "base/strings/string_piece_forward.h"
+#include "base/timer/timer.h"
+#include "brave/components/p3a/metric_log_store.h"
+#include "brave/components/p3a/metric_log_type.h"
+#include "brave/components/p3a/p3a_message.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+class PrefService;
+
+namespace network {
+class SharedURLLoaderFactory;
+}
+
+namespace p3a {
+
+struct P3AConfig;
+
+class Uploader;
+class StarLogStore;
+class RotationScheduler;
+class Scheduler;
+class StarHelper;
+class Uploader;
+
+struct RandomnessServerInfo;
+
+// The message manager has multiple roles related to handling/reporting metric
+// values. Metric updates received upstream from the Service are stored
+// in their appropriate LogStore instances. The Scheduler calls
+// methods in this class via callbacks and propagates metric upload to the
+// Uploader. The RotationScheduler also calls methods in this
+// class to handle reporting period rotation. STAR message preparation is also
+// triggered from this class.
+class MessageManager : public MetricLogStore::Delegate {
+ public:
+  using IsDynamicMetricRegisteredCallback =
+      base::RepeatingCallback<bool(const std::string& histogram_name)>;
+  class Delegate {
+   public:
+    virtual absl::optional<MetricLogType> GetDynamicMetricLogType(
+        const std::string& histogram_name) const = 0;
+    virtual void OnRotation(MetricLogType log_type, bool is_star) = 0;
+    // A metric "cycle" is a transmission to the P3A JSON server,
+    // or a STAR preparation for the current epoch.
+    virtual void OnMetricCycled(const std::string& histogram_name,
+                                bool is_star) = 0;
+    virtual ~Delegate() {}
+  };
+  MessageManager(PrefService* local_state,
```

Comment:
`= delete` copy and assignment, pls check all the classes in this pr follow the CC